### PR TITLE
[lldb/Commands] Add `scripting template list` command with auto discovery

### PR DESCRIPTION
**Pull request by medismailben**
This patch introduces a new `template` multiword sub-command to the
`scripting` top-level command. As the name suggests, this sub-command
operates on scripting templates, and currently has the ability to
automatically discover the various scripting extensions that lldb
supports.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

### CODE REVIEW

**Comment by medismailben** [comment]
Depends on #97263

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (bulbazord) [inline lldb/source/Commands/CommandObjectScripting.cpp:211]: typo: `validy` -> `validity`?


Also, what's the use case for the check? I don't see you use it.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (bulbazord) [inline lldb/source/Commands/CommandObjectScripting.cpp:208]: Suggestion: `Available scripting templates`. My reason is that the command is `scripting template list` so I would expect it to tell me what the available scripting templates are.

Additionally, it might be useful to specify the language. So something like `Available scripting templates (Python)` or something like that. This would let you drop the `print_field("Language", "Python")` below.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (bulbazord) [inline lldb/source/Commands/CommandObjectScripting.cpp:243]: Increment `i` here instead of below? It would be more resilient to making mistakes when modifying this loop for any reason.

File: lldb/source/Plugins/CMakeLists.txt
Comment (bulbazord) [inline lldb/source/Plugins/CMakeLists.txt:33]: I don't think there's a benefit to splitting the plugins into two different global properties. Is there a reason why we should treat these script plugins differently? 

Also I see you doing that for the python plugins below, how about the Lua ones?

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (medismailben) [reply lldb/source/Commands/CommandObjectScripting.cpp:211]: It's used in the next line. If set, it will check that the value argument isn't empty.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (medismailben) [reply lldb/source/Commands/CommandObjectScripting.cpp:208]: I was thinking `scripting extensions` rather than `scripting templates` may be we can do both `scripting extension templates` here ?

For the second suggestion, the command has an optional `-l|--language  <script-language>` option to only print extension information for that specific language.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (jimingham) [inline lldb/source/Commands/CommandObjectScripting.cpp:183]: should be m_language, right?

Comment (medismailben) [reply lldb/source/Commands/CommandObjectScripting.cpp:183]: Correct

**Comment by medismailben** [comment]
FWIW, this is the output it's producing for now:
```
(lldb) scripting template list 
Available scripted affordances:
  Name: ScriptedProcessPythonInterface
  Language: Python
  Description: Mock process state
  Command Interpreter Usages:
    process attach -C <script-name> [-k key -v value ...]
    process launch -C <script-name> [-k key -v value ...]
  API Usages:
    SBAttachInfo.SetScriptedProcessClassName
    SBAttachInfo.SetScriptedProcessDictionary
    SBTarget.Attach
    SBLaunchInfo.SetScriptedProcessClassName
    SBLaunchInfo.SetScriptedProcessDictionary
    SBTarget.Launch

  Name: ScriptedThreadPlanPythonInterface
  Language: Python
  Description: Alter thread stepping logic and stop reason
  Command Interpreter Usages: thread step-scripted -C <script-name> [-k key -v value ...]
  API Usages: SBThread.StepUsingScriptedThreadPlan
```

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (jimingham) [inline lldb/source/Commands/CommandObjectScripting.cpp:193]: Given that the vast majority of our users use Python as their scripting language, could we default to that rather than None?

Comment (jimingham) [reply lldb/source/Commands/CommandObjectScripting.cpp:193]: Or maybe better eScriptLanguageDefault?

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (jimingham) [inline lldb/source/Commands/CommandObjectScripting.cpp:257]: Use ScriptInterpreter::LanguageToString rather than hard coding the values here.

File: lldb/source/Commands/Options.td
Comment (jimingham) [inline lldb/source/Commands/Options.td:847]: specific -> specified

Comment (jimingham) [reply lldb/source/Commands/Options.td:847]: I don't think that's what is currently implemented, however, since you reset the language to None, and then error out if it is...

File: lldb/source/Plugins/CMakeLists.txt
Comment (medismailben) [reply lldb/source/Plugins/CMakeLists.txt:33]: I tried reusing `LLDB_ALL_PLUGINS` in `lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt` but I couldn't make it work: Either the scripted interface plugins wouldn't register or we would fail to link.

Not all the Python interfaces are registered so far, same for Lua, they will be added over time.

I'll leave it as is for now but feel free to change this after it lands :)

**Review by jimingham — APPROVED** [review]
LGTM

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (JDevlieghere) [inline lldb/source/Commands/CommandObjectScripting.cpp:143]: `affordances templates/affordance templates`

File: lldb/include/lldb/Core/PluginManager.h
Comment (JDevlieghere) [inline lldb/include/lldb/Core/PluginManager.h:496]: Could the usages be their own objects? It would simplif the plugin API and if you add a dump method you could simplify the command object too.

File: lldb/include/lldb/Core/PluginManager.h
Comment (JDevlieghere) [inline lldb/include/lldb/Core/PluginManager.h:501]: This looks unused?

File: lldb/include/lldb/Interpreter/Interfaces/ScriptedInterface.h
Comment (JDevlieghere) [inline lldb/include/lldb/Interpreter/Interfaces/ScriptedInterface.h:76]: Without `GetScriptedInterfaceCreateCallbackAtIndex` I don't think you need this either.

File: lldb/include/lldb/lldb-private-interfaces.h
Comment (JDevlieghere) [inline lldb/include/lldb/lldb-private-interfaces.h:130]: Nor this.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (JDevlieghere) [inline lldb/source/Commands/CommandObjectScripting.cpp:220]: This could be part of the class I was suggestion earlier.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (JDevlieghere) [inline lldb/source/Commands/CommandObjectScripting.cpp:193]: What's `check_validity`? It doesn't look like it's ever set.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (JDevlieghere) [inline lldb/source/Commands/CommandObjectScripting.cpp:267]: Remove

File: lldb/source/Plugins/CMakeLists.txt
Comment (JDevlieghere) [reply lldb/source/Plugins/CMakeLists.txt:33]: What exactly didn't work? I can't assess if this workaround makes sense without understanding the problem. It's totally fine for some of those interfaces not to be plugins.

Comment (bulbazord) [reply lldb/source/Plugins/CMakeLists.txt:33]: I see why this doesn't work now. LLDB plugins are added to the `LLDB_PLUGINS` global property via the PLUGIN argument to `add_lldb_library`. You aren't adding the PLUGIN argument to the `add_lldb_library` invocation in `lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt`. This is how the build system knows the list of plugins.

But even if you were to remove this and add that `PLUGIN` argument, this still wouldn't work. I don't think we've explicitly documented this behavior, but LLDB expects each library to be one plugin, not three as you've done here. When LLDB creates a plugin, it's going to generate a `Plugins.def` file that LLDB uses to initialize and teardown all plugins. It has a list of entries that look like this:
`LLDB_PLUGIN(lldbPluginFooBar)`. There are some exceptions for the script interpreter plugins (not sure why, didn't check). It also expects the `LLDB_PLUGIN_DEFINE` argument to match the library name exactly (without the initial `lldbPlugin`, specifically).

The way I see it, you have three options:
1. Restructure the build so that there are 3 libraries, each one of them a plugin.
2. Change LLDB's CMake machinery to accept libraries that define multiple plugin interfaces. You'll need to modify `add_lldb_library` so that you can pass it a list of plugin interface names instead of assuming the plugin names matches the library's name.
3. Keep the workaround and document why it exists.

I would suggest (1) or (2). I think (3) is just kicking the can down the road. This is a solution that explicitly works around LLDB's existing plugin system, is difficult to discover without knowing its there, and could make future refactors more difficult.

File: lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt
Comment (medismailben) [inline lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt:25]: @bulbazord aren't we making 3 separate libraries here ?

Comment (bulbazord) [reply lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/CMakeLists.txt:25]: No, you're just adding some elements to the list `LLDB_EXTRA_SCRIPT_PLUGINS`. Right below this is `add_lldb_library` which creates one library. It will contain the 3 cpp files that implement these plugin interfaces.

File: lldb/include/lldb/Target/DynamicLoader.h
Comment (bulbazord) [inline lldb/include/lldb/Target/DynamicLoader.h:336]: You'll probably need to rebase this change right?

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (bulbazord) [inline lldb/source/Commands/CommandObjectScripting.cpp:162]: Suggestion: `SetErrorStringWithFormatv` to avoid the `.str().c_str()` dance

**Review by bulbazord — COMMENTED** [review]
Seems alright to me, left a few minor suggestions inline.

**Review by bulbazord — APPROVED** [review]
LGTM, you'll probably want sign off from Jonas too though since he had a ton of comments.

File: lldb/source/Commands/CommandObjectScripting.cpp
Comment (JDevlieghere) [reply lldb/source/Commands/CommandObjectScripting.cpp:190]: You marked this as resolved because you check its value, but that variable is still never set?

**Review by JDevlieghere — APPROVED** [review]
LGTM modulo comment.

**Comment by mysterymath** [comment]
The reland broke the Fuchsia Windows build with a similar error to the LLVM Windows bots, but I wasn't able to cleanly revert this commit without detailed knowledge of the contents. Could you revert this (preferentially) or fix forward?

**Comment by medismailben** [comment]
> The reland broke the Fuchsia Windows build with a similar error to the LLVM Windows bots, but I wasn't able to cleanly revert this commit without detailed knowledge of the contents. Could you revert this (preferentially) or fix forward?
> 
> https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-windows-x64/b8741458989697976961/overview
> 
> https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8741458989697976961/+/u/clang/build/stdout

Looking at this. Thanks

**Comment by zeroomega** [comment]
Hi @medismailben 
We are seeing redefinition of pid_t error again on windows builder: https://lab.llvm.org/buildbot/#/builders/141/builds/1254
after multiple revert and reland like e72cdae47b4e263ea97b2bdd75cf44c1510cf3be. Could you revert the change please? Could you also not testing code in production please? It is a bad practice.

**Comment by DavidSpickett** [comment]
I thought I saw that fixed once already, but perhaps that was just a full revert.

We (Linaro) can test a patch for you if it's fiddly to verify, we just need to know the general idea of it.

**Comment by DavidSpickett** [comment]
Actually I'm looking at old results, the latest build makes it to tests https://lab.llvm.org/buildbot/#/builders/141/builds/1285 (which fail for reasons unrelated to this patch).

**Comment by medismailben** [comment]
Hey @DavidSpickett! Thanks for the help! I don't have a system to try this on and it only fails on windows :/

> Actually I'm looking at old results, the latest build makes it to tests https://lab.llvm.org/buildbot/#/builders/141/builds/1285 (which fail for reasons unrelated to this patch).

My patch has already been reverted in 9effefbae8d96006a4dd29bb9ab8532fd408559d and the failure you're talking about was introduced in #100710. I believe @Michael137 is trying to fix it.

If you have a recent build of lldb on windows (with python enabled), could you run the `scripting template list` command and paste the output here.

I've split this patch in smaller chunks (bccff3baeff8, ecf125eaf0b1, 995f643f5aa7) and they've all landed without any build failure on windows except for 2402b3213c2f. I've noticed however that the cmake configure phase would produce some warnings on the bot regarding the files related to my changes, about the fact that they'd exceed `CMAKE_OBJECT_PATH_MAX` limit (250 on windows apparently).

You can see it happening here : https://lab.llvm.org/buildbot/#/builders/141/builds/1222

```
CMake Warning in C:/Users/tcwg/llvm-worker/lldb-aarch64-windows/llvm-project/lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedProcessPythonInterface/CMakeLists.txt:
  The object file directory
    C:/Users/tcwg/llvm-worker/lldb-aarch64-windows/build/tools/lldb/source/Plugins/ScriptInterpreter/Python/Interfaces/ScriptedProcessPythonInterface/CMakeFiles/lldbPluginScriptInterpreterPythonScriptedProcessPythonInterface.dir/./
  has 227 characters.  The maximum full path to an object file is 250
  characters (see CMAKE_OBJECT_PATH_MAX).  Object file
    ScriptedProcessPythonInterface.cpp.obj
  cannot be safely placed under this directory.  The build may not work
  correctly.
```

Also, if you could re-apply e72cdae47b4e263ea97b2bdd75cf44c1510cf3be and attach your build log here, that would be very helpful, and hopefully help us narrow down the ambiguous typedef issue. 

If I'm not able to make it work on windows, I'll unfortunately have to compile it out of lldb for that platform, since I don't have access to a machine to investigate the issue further.

Thank you in advance!

**Comment by DavidSpickett** [comment]
I got it to build - https://github.com/llvm/llvm-project/pull/101672.

**Pull request merged**
